crates/sim1h/src/dht/bbdht/dynamodb/api/item/write.rs

```diff
@@ -27,27 +25,147 @@ pub fn ensure_content(
         String::from(CONTENT_KEY),
         string_attribute_value(&String::from(content.content())),
     );
+    item
+}
 
-    client
-        .put_item(PutItemInput {
-            item: item,
-            table_name: table_name.to_string(),
-            ..Default::default()
-        })
-        .sync()?;
-    Ok(())
+pub fn should_put_item_retry(
+    log_context: &LogContext,
+    put_item_result: Result<PutItemOutput, RusotoError<PutItemError>>,
+) -> BbDhtResult<bool> {
+    match put_item_result {
+        // no need to retry any success
+        Ok(_) => Ok(false),
+        Err(RusotoError::Service(err)) => match err {
+            PutItemError::InternalServerError(err) => {
+                // retry InternalServerErrors as these often seem to be temporary
+                tracer(
+                    &log_context,
+                    &format!("retry Service InternalServerError {:?}", err),
+                );
+                Ok(true)
+            }
+            PutItemError::ProvisionedThroughputExceeded(err) => {
+                // retry throughput issues as these will hopefully recover
+                tracer(
+                    &log_context,
+                    &format!("retry Service ProvisionedThroughputExceeded {:?}", err),
+                );
+                Ok(true)
+            }
+            PutItemError::RequestLimitExceeded(err) => {
+                // retry request limits as these will hopefully recover
+                tracer(
+                    &log_context,
+                    &format!("retry put_aspect Service RequestLimitExceeded {:?}", err),
+                );
+                Ok(true)
+            }
+            PutItemError::TransactionConflict(err) => {
+                // retry transaction conflicts
+                tracer(
+                    &log_context,
+                    &format!("retry Service TransactionConflict {:?}", err),
+                );
+                Ok(true)
+            }
+            // forward other put item errors back up the stack
+            _ => Err(err.into()),
+        },
+        Err(RusotoError::Unknown(err)) => {
+            // retry anything we don't know about
+            tracer(&log_context, &format!("retry Unknown {:?}", err));
+            Ok(true)
+        }
+        // forward other errors back up the stack
+        Err(err) => Err(err.into()),
+    }
+}
```

Comment:
I'm guessing that this is just the MVP approach, and we can add things like exponential backoff later -- since with enough traffic this could e.g. create a positive feedback loop of throughput errors

Reply:
@maackle sure, the first person to break it with traffic can submit a patch :P